File: dpnp/l2_distance/CPU/l2_distance.py
# Copyright (C) 2017-2018 Intel Corporation
#
# SPDX-License-Identifier: MIT

import dpctl
import base_l2_distance
import dpnp as np

def l2_distance(a,b):
    sub = a-b
    sq = np.square(sub)
    sum = np.sum(sq)
    d = np.sqrt(sum)
    return d

def l2_distance_dpctl(a,b):
    with dpctl.device_context("opencl:cpu"):
        l2_distance(a,b)

base_l2_distance.run("l2 distance", l2_distance_dpctl)

File: native/l2_distance/CPU/l2_distance.cpp
/*
 * Copyright (C) 2014-2015, 2018 Intel Corporation
 *
 * SPDX-License-Identifier: MIT
 */

#include <omp.h>
#include <mathimf.h>
#include "constants_header.h"

#ifdef __DO_FLOAT__
#   define SQRT(x)     sqrtf(x)
#else
#   define SQRT(x)     sqrt(x)
#endif

void l2_distance( int nopt, tfloat * x1, tfloat * x2, tfloat* distance_op ) {
  int i;
  tfloat a, b, sum = 0;

#pragma omp parallel for simd shared(x1, x2) reduction(+: sum)
  for ( i = 0; i < nopt; i++ ) {
    a = x1[i] - x2[i];
    b = a * a;
    sum += b;
  }
  *distance_op = SQRT(sum);
}
